README.md
#  Clipboard

Clipboard manager for macOS. It saves clipboard history and when you need to paste element which you copied long time ago, just press Commnad+Shift+V to active window and select element in history or multiple of them and paste by just pressing Return.

## Screenshots

<p align="center"><img src="https://raw.githubusercontent.com/paul-nameless/clipboard/master/screenshots/main_window.png"></p>
<p align="center"><img src="https://raw.githubusercontent.com/paul-nameless/clipboard/master/screenshots/screenshot.png"></p>

## Install

Just go to releases tab (https://github.com/paul-nameless/clipboard/releases) and download latest build or build it yourself in xcode, it is very easy.

## Shortcuts

Here are usefull shortcuts to know: 

* `Command+Shift+V` - show/hide clipboard history window
* `Return` - paste selected test in the row. If multiple rows selected will paste them separated with one space " "
* `Option+Return` - if multiple rows selected will join them and paste separated with no space
* `Cmd+Return` - if multiple rows selected will join them and paste separated with new line character "\n"
* `Delete` - delete selected text. If multiple selected will delete them all
* `Option + arrow down` - go to the last row
* `Option + arrow up` - go to the first row

## Build

You need to install Xcode firstly.

```
git clone https://github.com/paul-nameless/clipboard.git
cd clipboard
xcodebuild -project Clipboard.xcodeproj -alltargets -configuration Release
```


CHANGELOG.md
Changelog
=========

1.0.0 (2018-06-10)
------------------
- Initial release
- Copy paste as usual
- Open / hide clipbord history with cmd+shift+v
- When clipboard history window is active use shortcut cmd+(0-9) to parse first 10 items
- Delete item with (you guess) delete button. Also you can delete multiple items, just select multiple of them
- New liine characeters `\n` replaces with `⏎` to see the full line
- Multiple paste (select with shift of cmd multiple items and press enter to paste) separated with space (cmd+return to separate with new line characters and cmd+option to join string)
- Press esc or cmd+shift+v to hide window
- Tray that can be used to quit application or paste from history


ClipHistory/AppDelegate.swift
//
//  AppDelegate.swift
//  ClipHistory
//
//  Created by paul on 5/31/18.
//  Copyright © 2018 Paul Nameless. All rights reserved.
//

import Cocoa
import Foundation
import Magnet

@NSApplicationMain
class AppDelegate: NSObject, NSApplicationDelegate {

    let statusItem = NSStatusBar.system.statusItem(withLength:NSStatusItem.squareLength)
    let KEY_V: UInt16 = 0x09
    var timer = Timer()
    var clipboardChangeCount: Int = 0
    var preferencesController: NSWindowController?
    var clipboardHistoryController: NSWindowController?
    var stack: ClipboardStack = ClipboardStack()

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        buildMenu()
        setupShortcut()
       
        timer = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(self.checkClipboard(_:)), userInfo: nil, repeats: true)

//        // get a global concurrent queue
//        let background = DispatchQueue.global()
//        // submit a task to the queue for background execution
//        background.async {
//            if let clip = self.clipboardContent() {
//                print("Clipboard = \(clip)")
//            }
//        }

    }

    func applicationWillTerminate(_ aNotification: Notification) {
        // Insert code here to tear down your application
    }

    func iHaveFocus() -> Bool {
        let win = NSApplication.shared.keyWindow
        if (win != nil) {
            let vc = win!.contentViewController as? ClipboardHistoryViewController
            if (vc != nil) {
                return true
            }
        }
        return false
    }
    
    @objc func hotKeyHandler() {
        if iHaveFocus() {
            NSApp.hide(nil)
        } else {
            showClipboardHistory()
        }
    }
    
    func setupShortcut() {
        // ⌘ + Shift + V
        guard let keyCombo = KeyCombo(keyCode: 9, cocoaModifiers: [.shift, .command]) else { return }
        let hotKey = HotKey(identifier: "CommandShiftV",
                            keyCombo: keyCombo,
                            target: self,
                            action: #selector(AppDelegate.hotKeyHandler))
        hotKey.register()
    }


    func clipboardContent() -> String? {
        return NSPasteboard.general.pasteboardItems?.first?.string(forType: .string)
    }

    @objc func checkClipboard(_ sender: Any?) {
        if self.clipboardChangeCount != NSPasteboard.general.changeCount {
            self.clipboardChangeCount = NSPasteboard.general.changeCount
            if let clipboard = clipboardContent() {
                if clipboard.count < 100_000_000 {
                    stack.push(clipboard)
                    buildMenu()
                }
            }
        }
    }

    // init
    func buildMenu() {
        if let button = statusItem.button {
            button.image = NSImage(named:NSImage.Name("StatusBarButtonImage"))
            //            button.action = #selector(printQuote(_:))
        }

        let menu = NSMenu()
        
        let clipboards = stack.items.sorted { a, b in a.value > b.value }.enumerated().filter { i, e in i < 10 }
        for (i, elem) in clipboards {
            let key = indexToButton(i)
        
            let menuItem = NSMenuItem(title: trunc(string: elem.key, length: 44, trailing: "..."),
                                      action: #selector(AppDelegate.pasteElem(_:)),
                                      keyEquivalent: key)
            menuItem.tag = i
            menu.addItem(menuItem)
        }
//        menu.addItem(NSMenuItem.separator())
//        menu.addItem(NSMenuItem(title: "Clipboard",
//                                action: #selector(AppDelegate.showClipboardHistoryHandler(_:)),
//                                keyEquivalent: "V"))

//        menu.addItem(NSMenuItem.separator())
//        menu.addItem(NSMenuItem(title: "Clear", action: #selector(AppDelegate.clear(_:)), keyEquivalent: ""))
//        menu.addItem(NSMenuItem(title: "Preferences...",
//                                action: #selector(AppDelegate.showPreferences(_:)), keyEquivalent: ","))
//        menu.addItem(NSMenuItem(title: "Clipboard", action: #selector(AppDelegate.showAbout(_:)), keyEquivalent: ""))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Quit Clipboard", action: #selector(NSApplication.terminate(_:)), keyEquivalent: "Q"))
        
        statusItem.menu = menu
    }
    
    // button
    @objc func clear(_ sender: Any?) {
    }
    
    // button
    @objc func showAbout(_ sender: Any?) {
    }
    
    // button
    @objc func showPreferences(_ sender: Any?) {
        if preferencesController == nil {
            let storyboard = NSStoryboard(name: NSStoryboard.Name(rawValue: "Preferences"), bundle: nil)
            preferencesController = storyboard.instantiateInitialController() as? NSWindowController
        }

        if preferencesController != nil {
            preferencesController!.showWindow(nil)
//            print("Show pref")
        }
    }
    
    // button
    @objc func showClipboardHistoryHandler(_ sender: Any?) {
        showClipboardHistory()
    }

    func showClipboardHistory() {
        if clipboardHistoryController == nil {
            let storyboard = NSStoryboard(name: NSStoryboard.Name(rawValue: "ClipboardHistory"), bundle: nil)
            clipboardHistoryController = storyboard.instantiateInitialController() as? ClipboardHistoryWindowController
//            clipboardHistoryController.stack = self.stack
        }
        
        if clipboardHistoryController    != nil {
//            print("Show hist")
            clipboardHistoryController!.showWindow(nil)
            clipboardHistoryController?.window?.orderFront(clipboardHistoryController)
            NSApp.activate(ignoringOtherApps: true)
        }
    }
    
    @objc func pasteElem(_ sender: Any?) {
        let menuItem = sender as! NSMenuItem
        let index = menuItem.tag
        pasteByIndex(index)
    }
    
    func pasteByIndex(_ index: Int) {
        if let content = stack.get(index: index) {
            let pb = NSPasteboard.init(name: NSPasteboard.Name.general)
            pb.clearContents()
            pb.string(forType: NSPasteboard.PasteboardType.string)
            pb.declareTypes([NSPasteboard.PasteboardType.string], owner: nil)
            pb.setString(content, forType: NSPasteboard.PasteboardType.string)
            self.send(self.KEY_V, useCommandFlag: true)
        }
    }

    func pasteContent(_ content: String) {
        let pb = NSPasteboard.init(name: NSPasteboard.Name.general)
        pb.clearContents()
        pb.string(forType: NSPasteboard.PasteboardType.string)
        pb.declareTypes([NSPasteboard.PasteboardType.string], owner: nil)
        pb.setString(content, forType: NSPasteboard.PasteboardType.string)
        self.send(self.KEY_V, useCommandFlag: true)

    }

    func send(_ keyCode: CGKeyCode, useCommandFlag: Bool) {
        let sourceRef = CGEventSource(stateID: .combinedSessionState)
        
        if sourceRef == nil {
            NSLog("FakeKey: No event source")
            return
        }
        
        let keyDownEvent = CGEvent(keyboardEventSource: sourceRef,
                                   virtualKey: keyCode,
                                   keyDown: true)
        if useCommandFlag {
            keyDownEvent?.flags = .maskCommand
        }
        
        let keyUpEvent = CGEvent(keyboardEventSource: sourceRef,
                                 virtualKey: keyCode,
                                 keyDown: false)
        
        keyDownEvent?.post(tap: .cghidEventTap)
        keyUpEvent?.post(tap: .cghidEventTap)
    }

}




struct ClipboardStack {
    
    var items: [String: Double] = [:]
    let maxCapacity = 100
    let KEY = "clipboardStack"
    
    init() {
        let stackObject = UserDefaults.standard.object(forKey: self.KEY)
        if let stack = stackObject as? [String: Double] {
            self.items = stack
        } else {
            self.items = [:]
        }
    }
    
    func save() {
        UserDefaults.standard.set(self.items, forKey: self.KEY)
    }
    
    mutating func trim() {
        if (items.count > maxCapacity) {
            for _ in 1...items.count - maxCapacity {
                let element = items.max { a, b in a.value > b.value }?.key
                if let item = element {
                    items.removeValue(forKey: item)
                }
            }
        }
    }
    
    mutating func push(_ item: String) {
        items[item] = Date().timeIntervalSince1970
        trim()
        save()
    }
    
    func last() -> String? {
        return items.max { a, b in a.value > b.value }?.key
    }
    
    func get(index: Int) -> String? {
        let sortedItems = items.sorted { a, b in a.value > b.value }
        if index < sortedItems.count {
            return sortedItems[index].key
        }
        return nil
    }
    
    func getSorted() -> [(key: String, value: Double)] {
        return items.sorted { a, b in a.value > b.value }
    }

    mutating func remove(_ index: Int) {
        let e = getSorted()[index]
        items.removeValue(forKey: e.key)
        save()
    }
}


// utils
func indexToButton(_ i: Int) -> String {
    return i < 9 ? "\(i + 1)": i == 9 ? "0" : ""
}

// utils
func trunc(string: String, length: Int, trailing: String = "") -> String {
    return ((string.count > length) ? string.prefix(length) + trailing : string).replacingOccurrences(of: "\n", with: "⏎")
}



// Utils
//extension Date {
//
//    func getElapsedInterval() -> String {
//
//        let interval = Calendar.current.dateComponents([.year, .month, .day], from: self, to: Date())
//
//        if let year = interval.year, year > 0 {
//            return year == 1 ? "\(year)" + " " + "year" :
//                "\(year)" + " " + "years"
//        } else if let month = interval.month, month > 0 {
//            return month == 1 ? "\(month)" + " " + "month" :
//                "\(month)" + " " + "months"
//        } else if let day = interval.day, day > 0 {
//            return day == 1 ? "\(day)" + " " + "day" :
//                "\(day)" + " " + "days"
//        } else {
//            return "a moment ago"
//
//        }
//
//    }
//}


ClipHistory/ViewController.swift
//
//  ViewController.swift
//  ClipHistory
//
//  Created by paul on 5/31/18.
//  Copyright © 2018 Paul Nameless. All rights reserved.
//

import Cocoa
import HotKey

class ViewController: NSViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }

}

extension ViewController {
    // MARK: Storyboard instantiation
    static func freshController() -> ViewController {
        //1.
        let storyboard = NSStoryboard(name: NSStoryboard.Name(rawValue: "Main"), bundle: nil)
        //2.
        let identifier = NSStoryboard.SceneIdentifier(rawValue: "ViewController")
        //3.
        guard let viewcontroller = storyboard.instantiateController(withIdentifier: identifier) as? ViewController else {
            fatalError("Why cant i find ViewController? - Check Main.storyboard")
        }
        return viewcontroller
    }
}


ClipboardHistoryWindow/ClipboardHistoryViewController.swift
//
//  ClipboardHistoryViewController.swift
//  ClipHistory
//
//  Created by paul on 6/2/18.
//  Copyright © 2018 Paul Nameless. All rights reserved.
//

import Cocoa
import Carbon.HIToolbox

class ClipboardHistoryViewController: NSViewController {

    var appDelegate: AppDelegate? = nil
    let mapKeyCodeToNumber: [UInt16: Int] = [
        0x12: 1,
        0x13: 2,
        0x14: 3,
        0x15: 4,
        0x17: 5,
        0x16: 6,
        0x1A: 7,
        0x1C: 8,
        0x19: 9,
        0x1D: 0
    ]
    @IBOutlet weak var scrollView: NSScrollView!
    @IBOutlet weak var tableView: NSTableView!

    override func viewDidLoad() {

        super.viewDidLoad()
        self.parent?.view.window?.title = self.title!

        self.preferredContentSize = NSMakeSize(self.view.frame.size.width, self.view.frame.size.height)
        self.view.layer?.cornerRadius = 24
        
        appDelegate = NSApplication.shared.delegate as? AppDelegate
        tableView.delegate = self
        tableView.dataSource = self
        tableView.rowHeight = 44
        scrollView.wantsLayer = true
        scrollView.layer?.cornerRadius = 10
    }

    override func viewDidAppear() {
        super.viewDidAppear()
        self.tableView.reloadData()
        let indexSet = IndexSet.init(integer: 0)
        tableView.selectRowIndexes(indexSet, byExtendingSelection: false)
        self.parent?.view.window?.title = self.title!
    }
    
    override func keyDown(with event: NSEvent) {
        switch event.keyCode {
        case 0x33:
            // delete
            if let history = appDelegate?.stack.getSorted() {
                let indexes = tableView.selectedRowIndexes
                let itemsToDelete = indexes.filter { $0 < history.count }.map { history[$0].key }
                for item in itemsToDelete {
                    appDelegate?.stack.items.removeValue(forKey: item)
                }
                self.tableView.reloadData()
                tableView.selectRowIndexes(IndexSet(integer: indexes.first ?? 0), byExtendingSelection: false)
            }
        case 0x24:
            // enter
            NSApp.hide(nil)
            if let history = appDelegate?.stack.getSorted() {
                let flags = event.modifierFlags
                var separator = " "
                if flags.contains(NSEvent.ModifierFlags.command) {
                    separator = "\n"
                } else if flags.contains(NSEvent.ModifierFlags.option) {
                    separator = ""
                }
                let clipboard = tableView.selectedRowIndexes.filter { $0 < history.count }.map { history[$0].key }.joined(separator: separator)
                appDelegate?.pasteContent(clipboard)
                return
            }
        case 0x35:
            // escape
            NSApp.hide(nil)
        case 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x19, 0x1A, 0x1C, 0x1D:
            // 0-9
            let flags = event.modifierFlags
            if flags.contains(NSEvent.ModifierFlags.command) {
                NSApp.hide(nil)
                let index = mapKeyCodeToNumber[event.keyCode]!
                appDelegate?.pasteByIndex(index)
            }
        default:
            return
        }
    }
}


extension ClipboardHistoryViewController: NSTableViewDataSource {
    
    func numberOfRows(in tableView: NSTableView) -> Int {
//        let appDelegate = NSApplication.shared.delegate as! AppDelegate
        return appDelegate?.stack.items.count ?? 0
    }
    
}


class MyNSTableRowView: NSTableRowView {
    
    // Used to change selection color and round corners
    override func drawSelection(in dirtyRect: NSRect) {
        self.wantsLayer = true
        self.layer?.cornerRadius = 16
        if self.selectionHighlightStyle != .none {
            let selectionRect = NSInsetRect(self.bounds, 2.5, 2.5)
            // gray
//            NSColor(calibratedWhite: 0.81, alpha: 1).setFill()
            // blue
            NSColor(calibratedRed: 0.0/255, green: 122.0/255, blue: 254.0/255, alpha: 1.0).setFill()
            let selectionPath = NSBezierPath.init(roundedRect: selectionRect, xRadius: 10, yRadius: 10)
            selectionPath.fill()
        }
    }
}


extension ClipboardHistoryViewController: NSTableViewDelegate {
    
    fileprivate enum CellId {
        static let clipboardCell = "clipboard"
        static let timestampCell = "timestamp"
        static let shortcutCell = "shortcut"
    }
    
    func tableView(_ tableView: NSTableView, rowViewForRow row: Int) -> NSTableRowView? {
        return MyNSTableRowView()
    }

//    func tableViewSelectionDidChange(_ notification: Notification) {
//        let index = tableView.selectedRow
//        let rowView = tableView.rowView(atRow: index, makeIfNecessary: false)
//        rowView?.isEmphasized = false
//    }

    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        
//        let c: NSTextFieldCell? = (tableColumn?.dataCell as? NSTextFieldCell)!
//        if let c2 = c {
//            print("Here")
//            c2.textColor = NSColor.black
//        }
        var text: String = ""
        var cellIdentifier: String = ""
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .long
        dateFormatter.timeStyle = .long

        guard let item = appDelegate?.stack.getSorted()[row] else {
            return nil
        }

        if tableColumn == tableView.tableColumns[0] {
            text = truncForWindow(item.key)
            cellIdentifier = CellId.clipboardCell
        } else if tableColumn == tableView.tableColumns[1] {
            text = toTimeAgo(Date(timeIntervalSince1970: item.value))
            cellIdentifier = CellId.timestampCell
        } else if tableColumn == tableView.tableColumns[2] {
            text = row < 9 ? "⌘\(row + 1)": row == 9 ? "⌘0" : ""
            cellIdentifier = CellId.shortcutCell
        }
        
        if let cell = tableView.makeView(withIdentifier: NSUserInterfaceItemIdentifier(rawValue: cellIdentifier),
                                         owner: nil) as? NSTableCellView {
            cell.textField?.stringValue = text
//            cell.textField?.textColor = NSColor.black
//            cell.backgroundStyle = .light
            
            return cell
        }
        return nil
    }
    
}



extension NSTextFieldCell {
//    override open func drawingRect(forBounds rect: NSRect) -> NSRect {
//        return NSRect(x: rect.minX + 10, y: rect.minY + 10, width: rect.width, height: 44)
//    }
}

// utils

func truncForWindow(_ string: String) -> String {
    var str = string
    if str.count > 1000 {
        str = str.prefix(1000) + ""
    }
    str = str.replacingOccurrences(of: "\n", with: "⏎")
    return str
    
}

func toTimeAgo(_ date: Date) -> String {
    let interval = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute], from: date, to: Date())
    if let year = interval.year, year > 0 {
        return year == 1 ? "\(year)" + " " + "year" :
            "\(year)" + " " + "years"
    } else if let month = interval.month, month > 0 {
        return month == 1 ? "\(month)" + " " + "month" :
            "\(month)" + " " + "months"
    } else if let day = interval.day, day > 0 {
        return day == 1 ? "\(day)" + " " + "day" :
            "\(day)" + " " + "days"
    } else if let hour = interval.hour, hour > 0 {
        return hour == 1 ? "\(hour)" + " " + "hour" :
            "\(hour)" + " " + "hours"
    } else if let minute = interval.minute, minute > 0 {
        return minute == 1 ? "\(minute)" + " " + "minute" :
            "\(minute)" + " " + "minutes"
    } else {
        return "a moment ago"
        
    }
}
